qiskit/primitives/backend_estimator_v2.py

```diff
@@ -162,7 +162,10 @@ def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
         pub_dict = defaultdict(list)
         # consolidate pubs with the same number of shots
         for i, pub in enumerate(pubs):
-            shots = int(math.ceil(1.0 / pub.precision**2))
+            data = self._preprocess_pub(pub)
+            bc_obs = data.observables
+            coeff_sq_sum = sum(coeff**2 for coeff in bc_obs[0].values())
```

Comment:
Why are we looking only at the first observable? Don't we want to look at the worst case or something? Eg, consider the case

```
observables = [{"XX": 1}, {"YY": 100}]
```

Since we only look at the first observable, we'll be way off for the second one.

Reply:
Eg you could try
```python
        for i, pub in enumerate(pubs):
            coeff_sq_sum = 0
            for obs in pub.observables.ravel():
                coeff_sq_sum = max(coeff_sq_sum, sum(coeff**2 for coeff in obs.values()))
            shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
            pub_dict[shots].append(i)
```

Reply:
I tried this and it worked, thanks.